Web Extensions: Add the ability for extensions to delete bookmark nodes using

akutira-umich opened the pull request:
#### 086f3606d1bf705580e3061d207bfee7a3f1c43b
<pre>
Web Extensions: Add the ability for extensions to delete bookmark nodes using
the Bookmarks API
<a href="https://rdar.apple.com/152506190">rdar://152506190</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=296905">https://bugs.webkit.org/show_bug.cgi?id=296905</a>

Reviewed by Brian Weinstein.

Added UIProcess and WebProcess implementation more bookmarks.remove() and
bookmarks.removeTree(). Added tests for these functions as well

* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
(WebKit::WebExtensionContext::bookmarksCreate):
(WebKit::WebExtensionContext::bookmarksGetTree):
(WebKit::WebExtensionContext::bookmarksGetSubTree):
(WebKit::WebExtensionContext::bookmarksGet):
(WebKit::WebExtensionContext::bookmarksGetChildren):
(WebKit::WebExtensionContext::bookmarksGetRecent):
(WebKit::WebExtensionContext::bookmarksRemove):
(WebKit::WebExtensionContext::bookmarksRemoveTree):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::WebExtensionAPIBookmarks::remove):
(WebKit::WebExtensionAPIBookmarks::removeTree):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(findBookmarkAndParentArrayInMockTree):
(TestWebKitAPI::WKWebExtensionAPIBookmarks::configureCreateBookmarkDelegate):
(TestWebKitAPI::WKWebExtensionAPIBookmarks::configureRemoveBookmarksDelegate):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIRemoveAndRemoveTree)):
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
(-[TestWebExtensionsDelegate _webExtensionController:removeBookmarkWithIdentifier:removeFolderWithChildren:forExtensionContext:completionHandler:]):

Canonical link: <a href="https://commits.webkit.org/298453@main">https://commits.webkit.org/298453@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3deab6c0aa1e4c43d0a5489e5d3a5ee93e4d74a6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115490 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35175 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25683 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121567 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66054 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117379 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35841 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43750 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87745 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118438 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28591 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103658 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68137 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27744 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21778 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65230 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97977 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21895 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124730 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42417 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31784 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96518 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42784 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99848 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96304 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24509 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41539 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19401 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38310 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42299 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47862 "Built successfully") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41798 "Built successfully") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45126 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43514 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
```diff
@@ -141,6 +141,16 @@ WK_API_AVAILABLE(macos(15.4), ios(18.4), visionos(2.4))
+ @param context The context within which the web extension is running.
+ @param completionHandler A block that must be called upon completion.
+ */
+- (void)_webExtensionController:(WKWebExtensionController * _Nonnull)controller removeBookmarkWithIdentifier:(NSString *)bookmarkIdentifier recursively:(BOOL)recursively forExtensionContext:(WKWebExtensionContext * _Nonnull)context completionHandler:(void (^)(NSError * _Nullable))completionHandler;
```
I think renaming `recursively` to something like `removeTree` would make this more clear.

b-weinstein replied:
Or something like `removeFolderWithChildren` - since that is the difference in behavior with `remove` vs `removeTree`. I do like combining both into one delegate method though.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -427,11 +427,61 @@
+        return;
+    }
 
+    NSString *bookmarkIdString = bookmarkId.createNSString().get();
```
Nit: `bookmarkIdentifier`?

b-weinstein replied:
Or what you have is fine as well.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -427,11 +427,61 @@
 
+    [controllerDelegate _webExtensionController:controllerWrapper removeBookmarkWithIdentifier:bookmarkIdString recursively:YES forExtensionContext:contextWrapper completionHandler:^(NSError *error) {
+        if (error) {
+            completionHandler(toWebExtensionError(apiName, nullString(), @"error was reported"));
```
Can we include the actual error here?

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -427,11 +427,61 @@
+    NSString *bookmarkIdString = bookmarkId.createNSString().get();
+    [controllerDelegate _webExtensionController:controllerWrapper removeBookmarkWithIdentifier:bookmarkIdString recursively:NO forExtensionContext:contextWrapper completionHandler:^(NSError *error) {
+        if (error) {
+            completionHandler(toWebExtensionError(apiName, nullString(), @"error was reported"));
```
Can we include the actual error here?

b-weinstein replied:
It might make sense to use `NSString *privacyPreservingDescription(NSError *);` in CocoaHelpers or `localizedDescription` on the error. We do something like this the extensions window code:

```
    [m_delegate closeForWebExtensionContext:m_extensionContext->wrapper() completionHandler:makeBlockPtr([protectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler)](NSError *error) mutable {
        if (error) {
            RELEASE_LOG_ERROR(Extensions, "Error for window close: %{public}@", privacyPreservingDescription(error));
            completionHandler(toWebExtensionError(apiName, nullString(), error.localizedDescription));
            return;
        }

        completionHandler({ });
    }).get()];
```

Without this, there would be no way for the developer to know the error.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -574,6 +629,45 @@ void configureGetBookmarksDelegate(TestWebExtensionManager *manager)
+        @"let bookmarkB = await browser.bookmarks.create({ parentId: folder2.id, title: 'Bookmark B', url: 'http://example.com/b' });",
+        @"let topLevelBookmark = await browser.bookmarks.create({ title: 'Top Level Bookmark', url: 'http://example.com/top' });",
+        @"let tree1 = await browser.bookmarks.getTree();",
+        @"browser.test.assertEq(2, tree1[0].children.length, 'folder1 should have 2 children');",
```
I think `folder1` is wrong in this logging. This is the root folder, right?

akutira-umich replied:
I believe this is correct. Folder 1 is the first child under the root. tree1 is the top level vector so tree[0] is supposed to represent folder1. So tree[0].children is folder1's children which should have 2 children folder2 and bookmarkA

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -574,6 +629,45 @@ void configureGetBookmarksDelegate(TestWebExtensionManager *manager)
+        @"await browser.bookmarks.remove(folder1.id);",
+        @"let tree4 = await browser.bookmarks.getTree();",
+        @"browser.test.assertEq(1, tree4.length, 'the top level now ONLY has the top level bookmark');",
+        @"browser.test.notifyPass();",
```
Do you want to test any errors in this test case? I'd like to see:

1) Calling `removeTree` on a leaf fails
2) Calling `remove` on a folder with children fails
3) Calling `removeTree` on an empty folder fails

You could also add tests for testing removal of a non-existent identifier.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -134,6 +134,23 @@ @interface TestBookmarksDelegate : NSObject <WKWebExtensionControllerDelegatePri
     return nil;
 }
 
+static NSMutableDictionary *findBookmarkAndParentArrayInMockTree(NSMutableArray *rootChildren, NSString *bookmarkId)
```
Nit: I think `rootChildren` could just be `children` - since due to the recursion it won't always be the root.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -134,6 +134,23 @@ @interface TestBookmarksDelegate : NSObject <WKWebExtensionControllerDelegatePri
 
+static NSMutableDictionary *findBookmarkAndParentArrayInMockTree(NSMutableArray *rootChildren, NSString *bookmarkId)
+{
+    if (!rootChildren)
```
Nit: `if (!rootChildren.count)` - we could return early if it's an empty folder as well.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -307,6 +329,39 @@ void configureGetBookmarksDelegate(TestWebExtensionManager *manager)
+                if ([foundBookmarkType isEqualToString:@"folder"]) {
+                    NSArray *folderChildren = foundBookmarkDict[@"children"];
+
+                    if (folderChildren && folderChildren.count > 0) {
```
Nit: This can just be `if (folderChildren.count)`. If `folderChildren` is nil, `folderChildren.count` is safe to access and will always return 0.

b-weinstein replied:
This still needs to be fixed.

b-weinstein reviewed (changes requested):
This looks good! My biggest issue is the testing could be a little more fleshed out, but a very other small changes requested in the code.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
```diff
@@ -427,11 +427,61 @@
+    if (![controllerDelegate respondsToSelector:@selector(_webExtensionController:removeBookmarkWithIdentifier:removeFolderWithChildren:forExtensionContext:completionHandler:)]) {
+        completionHandler(toWebExtensionError(apiName, nullString(), @"it is not implemented"));
+        return;
+    }
```
Nit: Extra blank line here.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -134,6 +134,23 @@ @interface TestBookmarksDelegate : NSObject <WKWebExtensionControllerDelegatePri
+static NSMutableDictionary *findBookmarkAndParentArrayInMockTree(NSMutableArray *children, NSString *bookmarkId)
+{
+    if (!children)
+        return [NSMutableDictionary dictionary];
```
Why does this return an empty dictionary instead of nil? I don't think this is correct. If you have this:

```
            NSMutableDictionary *found = findBookmarkAndParentArrayInMockTree(childDict[@"children"], bookmarkId);
            if (found)
                return found;
```

`found` will always be non-nil if we hit the `!children` case.

akutira-umich replied:
Sorry this was to fix something from an early comment you made but I misinterpreted the line.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
```diff
@@ -574,6 +634,61 @@ void configureGetBookmarksDelegate(TestWebExtensionManager *manager)
+        @"browser.test.assertEq(1, tree2[0].children.length, 'folder1 should now have 1 child after removing bookmarkA');",
+        @"browser.test.assertEq('Folder Two', tree2[0].children[0].title, 'folder2 should be the only child of folder1');",
+        @"await browser.bookmarks.removeTree(folder2.id);",
+        @"let tree3 = await browser.bookmarks.getTree();",
```
Thanks for adding all of these tests. I think this is a big enough block of text that it could use some blank lines between test cases for readability (assuming that still compiles).

b-weinstein replied:
Like, especially around lines 650-652, where it's one line of code that is three lines of one API call.

akutira-umich committed:
Web Extensions: Add the ability for extensions to delete bookmark nodes using
the Bookmarks API
rdar://152506190
https://bugs.webkit.org/show_bug.cgi?id=296905

Reviewed by Brian Weinstein.

Added UIProcess and WebProcess implementation more bookmarks.remove() and
bookmarks.removeTree(). Added tests for these functions as well

* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionControllerDelegatePrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIBookmarksCocoa.mm:
(WebKit::WebExtensionContext::bookmarksCreate):
(WebKit::WebExtensionContext::bookmarksGetTree):
(WebKit::WebExtensionContext::bookmarksGetSubTree):
(WebKit::WebExtensionContext::bookmarksGet):
(WebKit::WebExtensionContext::bookmarksGetChildren):
(WebKit::WebExtensionContext::bookmarksGetRecent):
(WebKit::WebExtensionContext::bookmarksRemove):
(WebKit::WebExtensionContext::bookmarksRemoveTree):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIBookmarksCocoa.mm:
(WebKit::WebExtensionAPIBookmarks::remove):
(WebKit::WebExtensionAPIBookmarks::removeTree):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIBookmarks.mm:
(findBookmarkAndParentArrayInMockTree):
(TestWebKitAPI::WKWebExtensionAPIBookmarks::configureCreateBookmarkDelegate):
(TestWebKitAPI::WKWebExtensionAPIBookmarks::configureRemoveBookmarksDelegate):
(TestWebKitAPI::TEST_F(WKWebExtensionAPIBookmarks, BookmarksAPIRemoveAndRemoveTree)):
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
(-[TestWebExtensionsDelegate _webExtensionController:removeBookmarkWithIdentifier:removeFolderWithChildren:forExtensionContext:completionHandler:]):

Canonical link: https://commits.webkit.org/298453@main
